### ISSUE TITLE
Can I use toString or toSQL on a batchInsert call

### ISSUE THREAD

**Issue by mcbergsma** [knex/knex#5952]
I'm using Knex to convert JSON data into SQL Insert clauses for a project, so I'm never actually connecting to a DB, just specifying the client. I'd like to use the `batchInsert` util to generate a text file I can share, but it fails acquiring a connection (expected). Is there a way to just get the raw SQL generated by each chunk, or should I write my own chunker and do it manually via inserts?

**Comment by mercmobily** [comment]
Hi there, thanks for the question. Sorry for the delay... I assume you have moved on. Reply here for completeness.

This is expected behaviour; `batchInsert` is an execution helper and does not produce a query builder. Use chunked `insert` builders and call `toSQL()`/`toString()` on those.

`batchInsert` opens a transaction and executes each chunk; it returns a thenable with `.returning()`/`.transacting()` but it is not a builder, so there is nothing to compile to SQL. See lib/execution/batch-insert.js:6 for the helper entrypoint and lib/knex-builder/make-knex.js:119-122 where the method is wired. Because it executes, a connection acquisition is required even if you only want SQL text.

To generate SQL offline, chunk your data and compile a standard `insert` per chunk:

```
const { chunk } = require('lodash');

function compileBatchInsert(knex, tableName, rows, chunkSize = 1000) {
  return chunk(rows, chunkSize).map((rowsChunk) => {
    const qb = knex(tableName).insert(rowsChunk);
    const compiled = qb.toSQL(); // or qb.toSQL().toNative() for driver placeholders
    return {
      sql: compiled.sql,
      bindings: compiled.bindings,
      text: qb.toString(), // interpolated for debugging only
    };
  });
}
```

Notes:

- `toSQL().toNative()` returns driver-flavoured placeholders (e.g. `$1..$n` for PostgreSQL) and bindings; `toString()` is for debugging only and should not be executed directly against a database.
- A single multi-row insert per chunk is generally preferred to many single-row inserts for both performance and readability.

In conclusion: `batchInsert` itself cannot be stringified; use chunked `insert` builders to obtain SQL text for each chunk without touching a connection.
